PrestaShop/PrestaShop
# Check at installation if PrestaShop version is the latest

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Check at installation if PrestaShop version is the latest. If not, show the user a form to allow him to download and install the latest version.
| Type?         | new feature
| Category?     | IN
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/FF-173
| How to test?  | See below

## How to test

It requires a specific zip archive to test: it must be a zip archive built using this PR but with a version number lower than the current PS version.

Ask @matks (Mathieu Ferment) to build a fake v1.7.4.1 for your tests.

Then, you can install this fake install. The form will appear and ask you whether you wish to install the latest version instead of the 1.7.4.1:
- perform a regular 1.7.4.1 install by saying "No, thanks"
- download & install the latest (should be 1.7.4.2) by saying "Yes, please!"

## Pull Request summary

This PR does 3 things:

> 1) Modify compile process for releases

The PR modifies the behavior of `tools/build/Library/InstallUnpacker/compile.php` and related files to allow:
- injection of the release number into the built index.php using a placeholder
- injection of php classes into the built index.php using to keep a single file

> 2) Introduce a form at the install step to allow users to install the latest version

This basically introduces a 2nd page for the install step. It is being used only if the Installer is able to detect its version is not the latest stable available.

> 3) If the user chooses to update its install version, replace current install by the latest one

This means:
- to download the zip archive of the latest version
- to unzip the archive
- to replace current install files by the latest install files

If this process cannot be carried out, user is able to resume its regular installation. If possible, the reason why it could not be carried out is displayed as an error message before allowing the user to resume.

## Known limitations and possible improvements

There are multiple `@todo` statements in the code as I can see multiple improvements but did not want to make this PR too big and too complex.
The most relevant ones:
- php classes contain the Prestashop license header. When a new release is built, php classes are being copied into final build file `index.php` along with the license header who then appears multiple times in the final file.
- downloaded zip archive is installed right away, without checking md5 checksums
- no unit tests

Please tell me if you think these are vital for this PR and I will implement them.


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/9401)
